src/WebJobs.Script.WebHost/Diagnostics/MetricsEventManager.cs

```diff
@@ -462,30 +460,32 @@ internal void Flush()
             /// </summary>
             private void RaiseFunctionMetricEvents()
             {
-                if (_runningFunctions.Count == 0)
+                if (_runningFunctions.IsEmpty)
                 {
                     return;
                 }
 
                 // We only need to raise events here for functions that aren't completed.
                 // Events are raised immediately for completed functions elsewhere.
-                FunctionStartedEvent[] runningFunctionsSnapshot = null;
-                lock (_runningFunctionsSyncLock)
+                // Loop through and prune any completed runs
+                foreach (var possiblyRunning in _runningFunctions)
                 {
-                    // effectively we're pruning all the completed invocations here
-                    _runningFunctions = _runningFunctions.Where(p => !p.Completed).ToList();
-
-                    // create a snapshot within the lock so we can enumerate below
-                    runningFunctionsSnapshot = _runningFunctions.ToArray();
+                    if (possiblyRunning.Value.Completed)
+                    {
+                        _runningFunctions.TryRemove(possiblyRunning.Key, out _);
+                    }
                 }
 
-                // we calculate concurrency here based on count, since these events are raised
+                // If not for the length, we could probably avoid this array allocation
+                var concurrency = _runningFunctions.Count;
```

Comment:
Doesn't count acquire all the locks? Since you have already iterated through the collection above why not simply count there and than use that as the concurrency value or does the value need to be "precise" at the moment when you take it possibly including things that have been added in the meantime? I guess since there is no atomicity between those two iterations anyway and you take a snapshot of count counting above would be sufficient